### PR TITLE
Asynchronous DNS interface

### PR DESCRIPTION
**Pull request by mikaleppanen**
Interface will be merged to mbed-os master in:
https://github.com/ARMmbed/mbed-os/pull/6847

Header for DNS interface is here (on to be merged branch):
https://github.com/ARMmbed/mbed-os/blob/feature-emac/features/netsocket/DNS.h

### CODE REVIEW

**Comment by mikaleppanen** [comment]
@SeppoTakalo @AnotherButler please review.

File: docs/reference/api/connectivity/networksocket/DNS.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
+## DNS
+
+The DNS provides an interface to do DNS host name resolutions.
```
Comment (SeppoTakalo) [inline docs/reference/api/connectivity/networksocket/DNS.md]: We probably need to invent some name for this API.
It sound a bit weird to tell "DNS provides interface for DNS"

File: docs/reference/api/connectivity/networksocket/DNS.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+### DNS class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/class_dns.html)
+
+### Usage
+
+To make a DNS host name resolution: 
+
+1. Instantiate and connect network interface.
+1. Call the `gethostbyname` function to resolve address.
```
Comment (SeppoTakalo) [inline docs/reference/api/connectivity/networksocket/DNS.md]: Use `()` for functions to make it even more visible. So `gethostbyname()`

File: docs/reference/api/connectivity/networksocket/DNS.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
+## DNS
+
+The DNS provides an interface to do DNS host name resolutions.
+
+### DNS class reference
```
[the remark this answers is not in the document]
Comment (mikaleppanen) [reply docs/reference/api/connectivity/networksocket/DNS.md]: Corrected.

File: docs/reference/api/connectivity/networksocket/DNS.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/class_dns.html)
+
+### Usage
+
```
[the remark this answers is not in the document]
Comment (mikaleppanen) [reply docs/reference/api/connectivity/networksocket/DNS.md]: Added.

File: docs/reference/api/connectivity/networksocket/DNS.md (hunk 1 above)
Comment (mikaleppanen) [reply docs/reference/api/connectivity/networksocket/DNS.md]: Name is now: DNS resolver

File: docs/reference/api/connectivity/networksocket/DNS.md (hunk 2 above)
Comment (mikaleppanen) [reply docs/reference/api/connectivity/networksocket/DNS.md]: Corrected.

File: docs/reference/api/connectivity/networksocket/DNS.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
+## DNS Resolver
+
+The DNS resolver provides an interface to do DNS host name resolutions.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/networksocket/DNS.md]: Please expand on when and why you'd want to use this API.

File: docs/reference/api/connectivity/networksocket/DNS.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+### Usage
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/networksocket/DNS.md]: This section feels a lot like a tutorial. Is there any way to rephrase the numbered steps to give more background reference material than a step-by-step guide?

File: docs/reference/api/connectivity/networksocket/DNS.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    // Disconnect network interface
+    eth.disconnect();
+}
+
+```
+
+### DNS resolver class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/class_dns.html)
+
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/networksocket/DNS.md]: To follow our template, we need a transcluded example right here. If we don't already have one, could you please create one/copy the one above into Mbed teams? @SeppoTakalo Could you please help with this if needed?

File: docs/reference/api/connectivity/networksocket/DNS.md (hunk 5 above)
Comment (mikaleppanen) [reply docs/reference/api/connectivity/networksocket/DNS.md]: Added more description about dsn.

File: docs/reference/api/connectivity/networksocket/DNS.md (hunk 6 above)
Comment (mikaleppanen) [reply docs/reference/api/connectivity/networksocket/DNS.md]: Added a new chapter about asynchronous interface.

File: docs/reference/api/connectivity/networksocket/DNS.md (hunk 7 above)
Comment (mikaleppanen) [reply docs/reference/api/connectivity/networksocket/DNS.md]: I'll add example. Is "DNS_GetHostbyNameAsync_Example" a good name for it (follow naming of https://os.mbed.com/teams/mbed_example/code/TCPSocket_ConnStateCb_Example)?

Comment (SeppoTakalo) [reply docs/reference/api/connectivity/networksocket/DNS.md]: I'm fine with the example and the name.

**Comment by AnotherButler** [comment]
The Doxy link and the example link aren't working for me. Should the Doxy be a .h file or a class reference? 

I tried looking for the class reference here: https://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/annotated.html

But I could only find this .h file: https://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/nsapi__dns_8h_source.html

Is that the one I should use?

**Comment by mikaleppanen** [comment]
Yes, the pull request for DNS was merged yesterday to master. How about the mbed_example application, do we need to wait for mbed 5.9 release before publishing that?

**Comment by adbridge** [comment]
@mikaleppanen  

> Yes, the pull request for DNS was merged yesterday to master. How about the mbed_example application, do we need to wait for mbed 5.9 release before publishing that?

If this is an mbed-os official example then it needs to go in as part of the 5.9 release!! Please let us know as we will have to raise a PR to get it into the CI!!

**Comment by mikaleppanen** [comment]
I'm meaning https://os.mbed.com/teams/mbed_example/code/DNS_GetHostbyNameAsync_Example that I'm referring in document. I have not yet published that.

**Comment by SeppoTakalo** [comment]
@mikaleppanen please publish it before this gets merged in. Otherwise the example link is broken.

**Comment by mikaleppanen** [comment]
Published.

**Pull request merged**